cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
+
+Simultaneous CID management is only expected to happen during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+
+## KVStore compatibility
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with both `identity-allocation-mode=kvstore` and `identity-allocation-mode=crd`. CRD being the default, the initial implementation will use the CRD allocation mode, and later expanded to other allocation modes.
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manages-identities`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
+
+
+## Long term plan
+
+The long term plan is to make Operator managing CIDs the default, and deprecate the part of Agent that is currently used to manage CIDs.
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for pod creation. The control plane side of network policies and other features relying on CIDs will not carry out any required changes during this time, causing them to stop working.
+
+
+## Upgrade
+
+Safe upgrade, without downtime, requires simultaneous CID management compatibility.
+
+If we upgrade the Operator from a version that is not managing CIDs to a version that is managing CIDs while Agents are also managing CIDs, there should be no negative impact on the system. Everything should work the same.
+
+
+## Downgrade
+
+Downgrade without downtime is possible only if Agent daemonset is downgraded first -- updated to an earlier version where Agent manages CIDs. Otherwise, if Operator is downgraded first, there will be a downtime lasting at least the amount of time needed for Agent daemonset to be downgraded after the Operator downgrade.
```

Comment:
I think that the text should be more rigorous about describing the exact steps for upgrades, including which components are managing CIDs, and the intended supported upgrade/downgrade paths.

This paragraph currently seems like it is talking about going from *only* operator managing CIDs to *only* agent, but it seems to me like you don't intend on supporting that path. At the same time, the "upgrade" section above talks about needing to run simultaneous CID management. Presumably if the user is running simultaneous CID management during upgrade, then they will still be running this after upgrade. Maybe then they should run that way for a while and then later on disable agent CID management, once they are sure they will no longer need to downgrade?

We could even facilitate this, for example by saying that v1.17 creates the option for CID management from operator; v1.18 enables it by default with simultaneous management, and v1.19 disables agent CID management by default. We don't have to do it this way, but this is an example.

Reply:
I should also reinforce: If a user cannot downgrade safely, then I generally consider that they can also not upgrade safely. A clean upgrade is reliant on the ability to roll back to a safe older setup if any issue is encountered.

Reply:
I'm still debating back and forth myself about whether it's worth having an explicit "simultaneous management" stage or not. When thinking about upgrades in general, can be useful to break the upgrade into multiple stages. Each stage can be more easily tested since the target configuration can be explicitly configured for each state rather than relying on when the configuration is applied and which operators/agents were restarted to apply the new configuration. I also think that this would fit more easily with the `rolloutPods` option which automatically restarts Agents when configuration changes.

On the other hand I think that if you are managing your upgrade more carefully in order to upgrade Operator first, it should be totally viable to upgrade directly from CIDs being managed only by the agents to only managed by the operator.

Reply:
It’s worth putting more thought and time into converging on a solution.

To me it makes sense to do the following:
- Have two configuration flags, one for Operator and one for Agent, to allow more control.
- Treat simultaneous CID management as a transitional state, because that’s not where we want to end up. Although it will be a fully valid state, that doesn’t degrade user experience (any of the points mentioned in your [comment above](https://github.com/cilium/design-cfps/pull/59#discussion_r1812908975)).
- Define an upgrade path, where the Operator’s configuration flag should be always enabled first, before the Agent’s configuration flag is disabled. It will allow users to stay in the transitional state.
- Define a downgrade path, where the Agent’s configuration flag should be always enabled again, before the Operator’s configuration flag is disabled.

We could also consider making Agents manage CIDs whenever Agent’s flag is enabled OR Operator’s flag is disabled, to ensure that there is always at least one of them managing CIDs.

Reply:
SGTM for most of the above.

> We could also consider making Agents manage CIDs whenever Agent’s flag is enabled OR Operator’s flag is disabled, to ensure that there is always at least one of them managing CIDs.

I see where you're coming from, but I think this will be very counter-intuitive and confusing for users. We can complain if both flags are disabled, but I wouldn't enable functionality based on another flag being disabled.